### DISCUSSION TITLE
Would it be possible to modify the code to convert the JSON metadata back into the binary version?

### DISCUSSION THREAD

**Discussion by Trainmaster9977** [ScanMountGoat/Smush-Material-Research#41] [Ideas]
Hello!

I'm still working on that custom shader, and have run into some things where it'd be a lot easier if I could just pass a particular texture (specifically, the output of previous shaders) to it.

And given the nushdb metadata is part of what determines that (along with nufxlb, though unsure if it applies here) if I modified that, I could potentially do just that.

However, that would require converting the nushdb metadata back into the binary- but given this repo already converts it to json, how easy would it be to modify it so it can convert a json back?

**Comment by ScanMountGoat** [comment]
The .nushdb files are one of the simpler SSBH formats. The shader binaries within each entry in the nushdb don't use SSBH types and only have the minimal amount of work done to read out the metadata and decompile shaders. 

You would need to reverse engineer more of the binary format before implementing exporting. I don't have any plans to work on it personally, but I can review pull requests or give feedback on potential changes. From my experience looking at it, I would not expect adding write support to be easy. The metadata itself is mostly mapped out, but the rather large header is not.

ssbh_lib (nushdb):
https://github.com/ultimate-research/ssbh_lib/blob/master/ssbh_lib/src/formats/shdr.rs

ssbh_data (compiled binary + metadata inside each shader):
https://github.com/ultimate-research/ssbh_lib/blob/master/ssbh_data/src/shdr_data.rs

**Reply by Trainmaster9977** [reply]
Oh- for the record, I’m mostly *just* interested in the metadata. Unless the header affects it in a way I don’t know- 

Speaking of, while I have no clue about the 2896 header, Carlos and I did find the documentation of header for the code itself. The one that starts at the 2896 offset- I can send it’s over after I have dinner if you’d like

**Reply by ScanMountGoat** [reply]
From my experience with other games, I would expect there to be pointers and lengths for the sections starting with 0x34127698 (4.v˜) and 0x78563412 (xV4). I don't know if any of the data is relevant for the edits you'll be making. It might be easier to duplicate an existing shader that has the inputs you need already and just modify the compiled code.

**Reply by Trainmaster9977** [reply]
Well the edits I'll be making is passing the well, current screen texture(?. basically what you layer it onto at the end) as an input to the shader.

And I don't think any other material shaders do that?

https://download.nvidia.com/open-gpu-doc/Shader-Program-Header/1/Shader-Program-Header.html also for the record

**Reply by ScanMountGoat** [reply]
The material's vertex shader is executed for each vertex in the model. The material's fragment shader is executed for each pixel on screen covered by the model's geometry. The execution counts are a little more complex in practice. Reading and writing to the same resource like a texture generally isn't supported by any graphics API without explicit synchronization. Parallel code doesn't run in a predictable order, so you need to be careful about the order of writes and reads. If you want to modify the screen texture, you would need to write to a new texture that stores the results. I would recommend finding another way to do it.
